lib/rouge/lexers/hlsl.rb

```diff
@@ -0,0 +1,166 @@
+          Process2DQuadTessFactorsMax Process2DQuadTessFactorsMin
+          ProcessIsolineTessFactors ProcessQuadTessFactorsAvg
+          ProcessQuadTessFactorsMax ProcessQuadTessFactorsMin
+          ProcessTriTessFactorsAvg ProcessTriTessFactorsMax
+          ProcessTriTessFactorsMin QuadReadLaneAt QuadSwapX QuadSwapY radians
+          rcp reflect refract reversebits round rsqrt saturate sign sin sincos
+          sinh smoothstep sqrt step tan tanh tex1D tex1D tex1Dbias tex1Dgrad
+          tex1Dlod tex1Dproj tex2D tex2D tex2Dbias tex2Dgrad tex2Dlod tex2Dproj
+          tex3D tex3D tex3Dbias tex3Dgrad tex3Dlod tex3Dproj texCUBE texCUBE
+          texCUBEbias texCUBEgrad texCUBElod texCUBEproj transpose trunc
+          WaveAllBitAnd WaveAllMax WaveAllMin WaveAllBitOr WaveAllBitXor
+          WaveAllEqual WaveAllProduct WaveAllSum WaveAllTrue WaveAnyTrue
+          WaveBallot WaveGetLaneCount WaveGetLaneIndex WaveGetOrderedIndex
+          WaveIsHelperLane WaveOnce WavePrefixProduct WavePrefixSum
+          WaveReadFirstLane WaveReadLaneAt
+
+          SV_CLIPDISTANCE SV_CLIPDISTANCE0 SV_CLIPDISTANCE1 SV_CULLDISTANCE
+          SV_CULLDISTANCE0 SV_CULLDISTANCE1 SV_COVERAGE SV_DEPTH
+          SV_DEPTHGREATEREQUAL SV_DEPTHLESSEQUAL SV_DISPATCHTHREADID
+          SV_DOMAINLOCATION SV_GROUPID SV_GROUPINDEX SV_GROUPTHREADID
+          SV_GSINSTANCEID SV_INNERCOVERAGE SV_INSIDETESSFACTOR SV_INSTANCEID
+          SV_ISFRONTFACE SV_OUTPUTCONTROLPOINTID SV_POSITION SV_PRIMITIVEID
+          SV_RENDERTARGETARRAYINDEX SV_SAMPLEINDEX SV_STENCILREF SV_TESSFACTOR
+          SV_VERTEXID SV_VIEWPORTARRAYINDEX
+
+          allow_uav_condition branch call domain earlydepthstencil fastopt
+          flatten forcecase instance loop maxtessfactor numthreads
+          outputcontrolpoints outputtopology partitioning patchconstantfunc
+          unroll
+
+          BINORMAL BINORMAL0 BINORMAL1 BINORMAL2 BINORMAL3 BINORMAL4
+          BLENDINDICES0 BLENDINDICES1 BLENDINDICES2 BLENDINDICES3 BLENDINDICES4
+          BLENDWEIGHT0 BLENDWEIGHT1 BLENDWEIGHT2 BLENDWEIGHT3 BLENDWEIGHT4 COLOR
+          COLOR0 COLOR1 COLOR2 COLOR3 COLOR4 NORMAL NORMAL0 NORMAL1 NORMAL2
+          NORMAL3 NORMAL4 POSITION POSITION0 POSITION1 POSITION2 POSITION3
+          POSITION4 POSITIONT PSIZE0 PSIZE1 PSIZE2 PSIZE3 PSIZE4 TANGENT
+          TANGENT0 TANGENT1 TANGENT2 TANGENT3 TANGENT4 TESSFACTOR0 TESSFACTOR1
+          TESSFACTOR2 TESSFACTOR3 TESSFACTOR4 TEXCOORD0 TEXCOORD1 TEXCOORD2
+          TEXCOORD3 TEXCOORD4
+
+          FOG PSIZE
+
+          VFACE VPOS
+
+          DEPTH0 DEPTH1 DEPTH2 DEPTH3 DEPTH4
+        )
+      end
+
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z_][a-zA-Z0-9_]*/
+
+      state :root do
+        mixin :expr_whitespace
+        rule %r(
+          ([\w*\s]+?[\s*])                  # return arguments
+          (#{id})                           # function name
+          (\s*\([^;]*?\)(?:\s*:\s+#{id})?)  # signature
```

Comment:
@mitjmcc The final `}` in a function definition was being lexed as an error in functions that contained a semantic. I could have used `prepend` to add a new rule to the start of the state that the HLSL lexer inherits from the C lexer but it seemed preferable to me to redefine the entire state given how short it is.